### ISSUE TITLE
Error in `repo.active_branch` if branch is in DETACHED HEAD state

### ISSUE THREAD

**Issue by cgi1** [gitpython-developers/GitPython#633]
Refers to #511 
Error in `repo.active_branch` if branch is in `DETACHED HEAD` state:

```
{TypeError}HEAD is a detached symbolic reference as it points to '0fa1f2a7d9dd9fb63fa775d2d97ad72f56f1d819'
```

The branch is `None` in `def _get_ref_info(cls, repo, ref_path):` because of 

```
# its a commit
        if repo.re_hexsha_only.match(tokens[0]):
            return (tokens[0], None)
```

`def active_branch(self)` could return `HEAD-{commit-hash}` instead?

**Comment by cgi1** [comment]
Looks similar to  #629

**Comment by Byron** [comment]
I am not sure it would be right to return a commit instead. The function is called `active_branch`, and I would assume people expect it to return a branch, never a commit. If `HEAD` is detached, there is no branch to return.
What do you think?

**Comment by cgi1** [comment]
Hm good point, I don´t have a pefect answer.. To return no branch seems to be ok.

**Comment by Byron** [comment]
I think returning no branch would break existing code which expects an exception. On the other hand, the exception is not documented, so in a way it's running under the radar anyway.

Maybe it could be adjusted to return None if there is no active branch instead, which would make the calling code a little less cumbersome and safer, as it could start to expect that.

Would you like to contribute a respective PR?

**Comment by peterbe** [comment]
I stumbled on this too. With submodules. 
My solution wasn't to try to rely on a `try: except: ...` but I made it always check out at least some branch first before it started switching to the branch I ultimately wanted to. I.e. *always* start with `repo.heads[config["master_branch"]].checkout()`

**Comment by 1oser5** [comment]
If you just want to know which is current branch, you can do it by this
`
import git

repo = git.Repo()
branch  = repo.head.reference.name
`

**Comment by stas00** [comment]
3 years later and this is still a problem :(

**Comment by Byron** [comment]
> I think returning no branch would break existing code which expects an exception. On the other hand, the exception is not documented, so in a way it's running under the radar anyway.
> 
> Maybe it could be adjusted to return None if there is no active branch instead, which would make the calling code a little less cumbersome and safer, as it could start to expect that.
> 
> Would you like to contribute a respective PR?

I believe this comment still holds.

**Comment by stas00** [comment]
Thank you for this follow up, @Byron

My suggestion for the best solution would be this: In order to keep backward compatibility (exception that is) adding a new  argument that allows for what OP suggested:

```
git.Repo(search_parent_directories=True, active_branch_flexible=False)
```
and then:

```
if active_branch_flexible:
    try:
        active_branch = repo.active_branch
    except:
        active_branch = 'DETACHED_' + repo.head.object.hexsha
```

I haven't looked at the code, so this is totally a pseudocode based on what I saw in this issue.

Of course, if that suggestion resonates, you'd know better what to call it, `active_branch_flexible` is just a quick idea ;)

Thank you.

**Comment by Byron** [comment]
I think at this point it's clear that the methods behaviour is surprising, so having alternatives seems valuable.

Thanks @stas00 for the suggestion, which certainly works as to remove the exception that keeps tripping people up.
Making the returned branch a string as proposed above could be just as dangerous though, as the type it will return has never been a string. Also I would hope to keep things more local and avoid adding another, very specialized configuration flag to the repository.

My proposition is to add a new method or property along the lines of `active_branch_or_none` with documentation explaining under which circumstances `None` would be expected.

**Comment by stas00** [comment]
Surely at the very least `active_branch_or_none` will remove the element of surprise.

In the particular situation it's used where it tripped me up (surprise indeed!), returning `None` is just as good as throwing an exception, since I still need to replace it with some branch-like string. So I'd ideally want a new method that returns some sort of a string - active branch or perhaps even sha if it's not available. i.e., may be `active_branch_or_sha`. But whatever you decide is good for me, now that I added a try/except block anyway.

Thank you for your thoughtful commentary, @Byron.

**Comment by Byron** [comment]
Indeed, ideally there is a way to not enforce any branching, but I don't believe that exists since there either is a branch, or there is none. With a method name that makes this clear there should be no surprise, but I invite you to experiment with styles and names and submit a PR with what works best for you.

It really would be great if you could be the one to put an end to this issue :), so imagine me cheering you on from the sidelines :).

**Comment by stas00** [comment]
Thank you for this invitation, but I will pass at the moment. I have never used `gitpython` and it was in someone else's code totally unrelated to what I was doing, so I had to remove that obstacle to proceed. That's how I ended up here. 

Should I start using the library and get to learn how it works, I'd be happy to submit a PR.

**Comment by Eoin-McMahon** [comment]
3.5 years later and this is still a problem :(

**Comment by Byron** [comment]
[This comment](https://github.com/gitpython-developers/GitPython/issues/633#issuecomment-710748523) might help.

**Comment by adnanmuttaleb** [comment]
I read the last comment, but it was 4 years ago, any new update on this?

**Comment by Byron** [comment]
Soon this issue is 8 years old!

With typing it would make sense to return `None` if there is no active branch.
That should be an easy fix and PRs are welcome.

**Comment by ssbarnea** [comment]
I think that one of the reasons why nobody proposed a fix for this one is the lack of type info.

If someone with more knowledge about library internals can give a hint it would be useful.

- Git status reports something like `HEAD detached at ...`, 
- Repo.active_branch raises `HEAD is a detached symbolic reference as it points to ...`, 
- .active_branch signature returns a `Head(Reference)`

No idea if it is possible to return a valid object in this case. Making it return None would clearly break a lot of consumers that assume that they always receive a Head object.

**Comment by Byron** [comment]
`gitoxide` always [returns a `Head`](https://docs.rs/gix/latest/gix/head/enum.Kind.html), and takes it from there. But then again, there are no exceptions in Rust so could never confuse the developer of the API into thinking it's a good idea to actually use it to communicate a completely none-exceptional case.
